src/interceptors/ClientRequest/MockHttpSocket.ts

```diff
@@ -0,0 +1,457 @@
+import net from 'node:net'
+import {
+  HTTPParser,
+  type RequestHeadersCompleteCallback,
+  type ResponseHeadersCompleteCallback,
+} from '_http_common'
+import { STATUS_CODES } from 'node:http'
+import { Readable } from 'node:stream'
+import { randomUUID } from 'node:crypto'
+import { invariant } from 'outvariant'
+import { MockSocket } from '../Socket/MockSocket'
+import type { NormalizedWriteArgs } from '../Socket/utils/normalizeWriteArgs'
+import { isPropertyAccessible } from '../../utils/isPropertyAccessible'
+import { baseUrlFromConnectionOptions } from '../Socket/utils/baseUrlFromConnectionOptions'
+import { parseRawHeaders } from '../Socket/utils/parseRawHeaders'
+import { getRawFetchHeaders } from '../../utils/getRawFetchHeaders'
+import { RESPONSE_STATUS_CODES_WITHOUT_BODY } from '../../utils/responseUtils'
+
+type HttpConnectionOptions = any
+
+export type MockHttpSocketRequestCallback = (args: {
+  requestId: string
+  request: Request
+  socket: MockHttpSocket
+}) => void
+
+export type MockHttpSocketResponseCallback = (args: {
+  requestId: string
+  request: Request
+  response: Response
+  isMockedResponse: boolean
+  socket: MockHttpSocket
+}) => void
+
+interface MockHttpSocketOptions {
+  connectionOptions: HttpConnectionOptions
+  createConnection: () => net.Socket
+  onRequest: MockHttpSocketRequestCallback
+  onResponse: MockHttpSocketResponseCallback
+}
+
+const kRequestId = Symbol('kRequestId')
+
+export class MockHttpSocket extends MockSocket {
+  private connectionOptions: HttpConnectionOptions
+  private createConnection: () => net.Socket
+  private baseUrl: URL
+
+  private onRequest: MockHttpSocketRequestCallback
+  private onResponse: MockHttpSocketResponseCallback
+
+  private writeBuffer: Array<NormalizedWriteArgs> = []
+  private request?: Request
+  private requestParser: HTTPParser<0>
+  private requestStream?: Readable
+  private shouldKeepAlive?: boolean
+
+  private responseType: 'mock' | 'bypassed' = 'bypassed'
+  private responseParser: HTTPParser<1>
+  private responseStream?: Readable
+
+  constructor(options: MockHttpSocketOptions) {
+    super({
+      write: (chunk, encoding, callback) => {
+        this.writeBuffer.push([chunk, encoding, callback])
```

Comment:
I think we need to call the callback after we "flush" the chunk to the OS.
Related to this problem: https://github.com/mswjs/interceptors/pull/515?notification_referrer_id=NT_kwDOAK7cMLM5Njk2NjM3NjYxOjExNDU5NjMy#issuecomment-1995549971

Reply:
We shouldn't be doing any of this since we are writing to the socket, unless I'm getting it wrong, of course. That's the point of relying on `net.Socket` under the hood 🤔 

@mikicho, can you please share a use case that illustrates how this manifests?

Reply:
@kettanaito The thing is that the native `write` function we override (or one of its derivatives) changes the internal `_writableState`, which is used later.

>  can you please share a use case that illustrates how this manifests?
I think this is the root cause of this: https://github.com/mswjs/interceptors/pull/515#issuecomment-1995549971

And here is a failing test for `@mswjs/interceptors`:

```js
it.only('should fire finish event', async () => {
  interceptor.on('request', async ({ request }) => {
    request.respondWith(new Response(null, { status: 200 }))
  })
  const finishSpy = vi.fn()
  const request = http.request(httpServer.http.url('/resource'), {
    method: 'POST',
  })

  request.on('finish', finishSpy)
  request.end()

  const { res } = await waitForClientRequest(request)
  expect(res.statusCode).toBe(200) 
  expect(finishSpy).toHaveBeenCalledTimes(1) // AssertionError: expected "spy" to be called 1 times, but got 0 times
})
```

Reply:
@mikicho, opened a fix in https://github.com/mswjs/interceptors/pull/542. Your test is passing now. Also added a few missing tests. We weren't calling write callbacks in the mocked response scenarios at all, like you've pointed out. Also found a bug around `.push()` and the `end` event.